### PR TITLE
kubeadm: fix the bug that bazel build does not respect hack/print-workspace-status.sh

### LINKED ISSUE #1669: Kubeadm bazel build does not respect hack/print-workspace-status.sh anymore
## Is this a BUG REPORT or FEATURE REQUEST?
/kind bug

## Versions
onmaster

## What happened?
I'm trying to compile kubeadm on master, and the kubeadm binary does not get the version as usual but instead gots a wrong GitVersion:"v0.0.0-master+$Format:%h$"
```
kubeadm version: &version.Info{Major:"", Minor:"", GitVersion:"v0.0.0-master+$Format:%h$", GitCommit:"$Format:%H$", GitTreeState:"", BuildDate:"1970-01-01T00:00:00Z", GoVersion:"go1.12.7", Compiler:"gc", Platform:"linux/amd64"}
```

## What you expected to happen?
kubeadm get the expected version, consistent with hack/print-workspace-status.sh

## How to reproduce it (as minimally and precisely as possible)?
create a buildversion file with the following content
```
export KUBE_GIT_MAJOR=1
export KUBE_GIT_MINOR=15
export KUBE_GIT_VERSION=v1.15.0
export KUBE_GIT_TREE_STATE=clean
export KUBE_GIT_COMMIT=d34db33
```
Set the build version
```
export KUBE_GIT_VERSION_FILE=$working_dir/buildversion
```
then check the version is properly set
```
hack/print-workspace-status.sh
```
then build kubeadm
```
bazel build //cmd/kubeadm
```
and finally check kubeadm got the right version with `kubeadm version`

## Anything else we need to know?
probably this is a regression introduced by https://github.com/kubernetes/kubernetes/pull/78929

/cc @SataQiu 
/cc @rosti 
/cc @yastij 
/priority critical-urgent
/milestone v1.16

### PR DESCRIPTION
**Pull request by SataQiu**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
kubeadm: fix the bug that bazel build does not respect hack/print-workspace-status.sh

**Which issue(s) this PR fixes**:

Fixes kubernetes/kubeadm#1669

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by SataQiu** [comment]
If we want to clean up the dependency in `cmd/kubeadm/BUILD`，we may need probably do what kubectl did and have a local copy of `version_x_defs`. 
Ref: 
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubectl/BUILD#L6
https://github.com/kubernetes/kubernetes/tree/master/pkg/kubectl/version
https://github.com/kubernetes/kubernetes/blob/master/hack/lib/version.sh#L162

File: cmd/kubeadm/BUILD [hunk 1]
```diff
@@ -5,7 +5,7 @@ load(
     "go_binary",
     "go_library",
 )
-load("//vendor/k8s.io/client-go/pkg/version:def.bzl", "version_x_defs")
```
Comment (rosti) [inline cmd/kubeadm/BUILD]: Let's see what the actual problem is, instead of reverting straight to this.
/hold

**Comment by rosti** [comment]
I am more inclined to just copy the `//pkg/version` variant of `version_x_defs` locally. This is what kubectl does. In fact, I was suspecting problems with client-go's version of it for a start.

**Comment by yastij** [comment]
I agree with @rosti we should have `version_x_defs ` locally as kubectl and client-go does

**Comment by fabriziopandini** [comment]
@SataQiu thanks for the quick fix.
Unfortunately the output change a little bit, but this is not resolutive
```
kubeadm version: &version.Info{Major:"{gitMajor}", Minor:"{gitMinor}", GitVersion:"{gitVersion}", GitCommit:"{gitCommit}", GitTreeState:"{gitTreeState}", BuildDate:"{buildDate}", GoVersion:"go1.12.7", Compiler:"gc", Platform:"darwin/amd64"}
```
@rosti @yastij someone can kindly double check to exclude this is a problem of my setup?

**Comment by rosti** [comment]
> @SataQiu thanks for the quick fix.
> Unfortunately the output changes a little bit, but this is not resolutive
> 
> ```
> kubeadm version: &version.Info{Major:"{gitMajor}", Minor:"{gitMinor}", GitVersion:"{gitVersion}", GitCommit:"{gitCommit}", GitTreeState:"{gitTreeState}", BuildDate:"{buildDate}", GoVersion:"go1.12.7", Compiler:"gc", Platform:"darwin/amd64"}
> ```
> 
> @rosti @yastij someone can kindly double check to exclude this is a problem of my setup?

Without this fix:
```
kubeadm version: &version.Info{Major:"", Minor:"", GitVersion:"v0.0.0-master+$Format:%h$", GitCommit:"$Format:%H$", GitTreeState:"", BuildDate:"1970-01-01T00:00:00Z", GoVersion:"go1.12.7", Compiler:"gc", Platform:"linux/amd64"}
```

With the fix:
```
kubeadm version: &version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.0-alpha.0.2252+303509cdf50c0b-dirty", GitCommit:"303509cdf50c0bc711e837654ac9fec5a06c6976", GitTreeState:"dirty", BuildDate:"2019-07-15T11:58:31Z", GoVersion:"go1.12.7", Compiler:"gc", Platform:"linux/amd64"}
```

Yet, this is odd, because we still use client-go's version package in the `kubeadm version` command.

I also tested a rough local to kubeadm copy of kubectl's version package and it works correctly.

**Comment by yastij** [comment]
@rosti - here w're referring to `vendor/k8s.io/client-go/pkg/version`'s def.bzl. it contains the following:

```
  stamp_pkgs = [
      "k8s.io/kubernetes/pkg/version",
      # In hack/lib/version.sh, this has a vendor/ prefix. That isn't needed here?
      "k8s.io/client-go/pkg/version",
      ]
```

I think it should be `"k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version"`

**Comment by fabriziopandini** [comment]
After reverting bazel from 0.27.0 to 0.25.1 I can confirm the fix from @SataQiu works also for me.

I let to @rosti and @yastij the last work on use this fix as it is or if to change as suggested.
My +1 goes to avoid the mix between client-go and pkg/version and be as much as possible consistent with kubectl/other commands

**Comment by rosti** [comment]
This is currently in a some sort of a middle state. We use both `//pkg/version` and `//vendor/k8s.io/client-go/pkg/version` at the moment. Also, client-go is versioned differently, so I am not sure if it can start giving us client-go version numbers.
Hence, I am in favor of having a local copy of this package somewhere. A quick and dirty POC patch for this can be found [here](https://github.com/rosti/kubernetes/commit/9e0c42d2904add019684175ed7395831769d9b5e). It's a misplaced copy of kubectl's version package.

**Comment by neolit123** [comment]
ok, let's have the local copy.

**Review by fejta — APPROVED** [review]
/approve
the hack change

**Comment by neolit123** [comment]
/assign @rosti
please approve the kubeadm change if you think this is good.

**Review by yastij — COMMENTED** [review]
/lgtm

**Review by rosti — COMMENTED** [review]
Thanks @SataQiu !
Looks superb, but can we move the version package to `//cmd/kubeadm/app/version`?

**Review by rosti — APPROVED** [review]
Thanks @SataQiu !
/approve
/lgtm
/hold cancel

**Pull request merged**